### ISSUE TITLE
Allow for custom AgroalDataSourceListener

### ISSUE THREAD

**Issue by segersb** [quarkusio/quarkus#36263]
### Description

I would like to be able to configure a custom AgroalDataSourceListener.

I found the following issue from 2019, but the change only allowed for a custom AgroalPoolInterceptor:
"Allow addition of AgroalDataSourceListener implementations during run-time production of DataSources."
https://github.com/quarkusio/quarkus/issues/5381

### Implementation ideas

I would propose to have a similar solution for custom AgroalDataSourceListener as there currently is for AgroalPoolInterceptor.

A custom AgroalPoolInterceptor can be defined by making an ApplicationScoped implementation of the interface.
It is also possible to annotate the interceptor class with the @DataSource qualifier.

The exact same setup could be done for custom AgroalDataSourceListener.
The change would have to be done in io.quarkus.agroal.runtime.DataSources.

```
private final Instance<AgroalPoolInterceptor> agroalPoolInterceptors; // existing
private final Instance<AgroalDataSourceListener > agroalDataSourceListeners; // to be added
```

Currently there is only 1 hard-coded listener for logging
This part should be adapted to include the custom agroalDataSourceListeners as well
```
AgroalDataSource dataSource = new io.agroal.pool.DataSource(agroalConfiguration,
        new AgroalEventLoggingListener(dataSourceName,
                agroalConfiguration.connectionPoolConfiguration()
                        .transactionRequirement() == TransactionRequirement.WARN));
```

**Comment by yrodiere** [comment]
Hi. Is this different from https://github.com/quarkusio/quarkus/pull/5408 ?

@barreiro , the person most likely to approve such a feature, had concern with the approach: https://github.com/quarkusio/quarkus/pull/5408#issuecomment-579857530
Do you have ideas to address those concerns?

**Comment by segersb** [comment]
All valid concerns, I guess it's the responsibility of the developer to know what they are doing.

The AgroalPoolInterceptor that can be used now also gives access to the underlying connection.
Maybe it has less room for errors, but I'd say there are infinite ways to mess up if not careful :)

For context, I need to make sure that connections are rolled back before they are closed.
Oracle seems to perform a commit when a connection is closed.
We have an issue when quarkus is shut down gracefully while there are long-running tasks running.
I created an issue for this specific problem as well: https://github.com/quarkusio/quarkus/issues/36265

**Comment by segersb** [comment]
Would it be possible to get an update on this, we currently have to resort to pretty ugly code to have a custom AgroalDataSourceListener so that transaction management for Oracle works properly under Quarkus.
See https://github.com/quarkusio/quarkus/issues/36265 for more details on that

**Comment by segersb** [comment]
@barreiro
Would it be possible to look at this issue?
We currently have to rely on a dirty workaround to have transaction integrity. This would allow us to have proper code.

Alternatively, Quarkus could have a built-in datasource listener that does the rollback before close. (needed for Oracle)
```
    @Override
    public void beforeConnectionDestroy(Connection connection) {
        try {
            connection.rollback();
        } catch (SQLException e) {
            // ignore
        }
    }
```

**Comment by barreiro** [comment]
@segersb Agroal 2.3 adds `onConnectionCreate` / `onConnectionDestroy` methods to `AgroalPoolInterceptor`. 

This will support you use case (if Agroal prevents the rollback, you'll need to unwrap the connection and call `rollback` on the raw connection).

This methods run on Agroal housekeeping thread and therefore should not be abused.

**Comment by segersb** [comment]
@barreiro
Thanks, that would already be a much better solution than what we have now!

Is that something that Quarkus would implement by default?
It doesn't seem right that transactions are not safe by default when using Oracle.

**Comment by barreiro** [comment]
@segersb fixed in #39072

**Comment by yrodiere** [comment]
Changed the title to clarify that the solution was not to allow custom `AgroalDataSourceListener` in the end, but this:

> @segersb Agroal 2.3 adds `onConnectionCreate` / `onConnectionDestroy` methods to `AgroalPoolInterceptor`.
> 
> This will support you use case (if Agroal prevents the rollback, you'll need to unwrap the connection and call `rollback` on the raw connection).
> 
> This methods run on Agroal housekeeping thread and therefore should not be abused.
